test/jdk/java/lang/reflect/code/writer/TestOpMethod.java

```diff
@@ -0,0 +1,163 @@
+        CLASS_MODEL = ClassFile.of().parse(IR.class.getResourceAsStream("IR.class").readAllBytes());
+    }
+
+    @DataProvider
+    byte[][] classes() {
+        try {
+            return new byte[][] {
+                    IR.class.getResourceAsStream("IR.class").readAllBytes()
+            };
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Test(dataProvider = "classes")
+    void test(byte[] cd) throws Throwable {
+        var cf = ClassFile.of();
+        var cm = cf.parse(cd);
+        var opFieldsAndIRs = getOpFieldsAndIRs(cm);
+        cd = removeOpFields(cm);
+        for (var opFieldAndIR : opFieldsAndIRs) {
+            var funcOp = ((FuncOp) OpParser.fromStringOfFuncOp(opFieldAndIR.ir()));
+            var builderOp = OpBuilder.createBuilderFunction(funcOp);
+            var builtOp = (Op) Interpreter.invoke(builderOp, ExtendedOp.FACTORY, CoreTypeFactory.CORE_TYPE_FACTORY);
+            Assert.assertEquals(builtOp.toText(), opFieldAndIR.ir());
+
+            cd = BytecodeGenerator.addOpByteCodeToClassFile(MethodHandles.lookup(), cm, opFieldAndIR.fieldName(), builderOp);
+            cm = cf.parse(cd);
+            Assert.assertEquals(cm.methods().stream().filter(mm -> mm.methodName().equalsString(opFieldAndIR.fieldName())).count(),
+                    opFieldsAndIRs.size());
+
+            var l = MethodHandles.lookup().defineHiddenClass(cd, true);
+            var mh = l.findStatic(l.lookupClass(),
+                    opFieldAndIR.fieldName(),
+                    MethodRef.toNominalDescriptor(builderOp.invokableType()).resolveConstantDesc(l));
+            Assert.assertEquals(
+                    ((Op) mh.invoke(ExtendedOp.FACTORY, CoreTypeFactory.CORE_TYPE_FACTORY)).toText(),
+                    opFieldAndIR.ir());
+        }
+        print(cm);
+    }
+
+    static void print(byte[] bytes) {
+        print(ClassFile.of().parse(bytes));
+    }
+
+    static void print(ClassModel cm) {
+        ClassPrinter.toYaml(cm, ClassPrinter.Verbosity.TRACE_ALL, System.out::print);
+    }
+
+    record OpFieldAndIR(String fieldName, String ir) {}
+
+    List<OpFieldAndIR> getOpFieldsAndIRs(ClassModel cm) {
+        List<OpFieldAndIR> res = new ArrayList<>();
+        var cinit = cm.methods().stream().filter(mm -> mm.methodName().equalsString(ConstantDescs.CLASS_INIT_NAME))
+                .findFirst().orElseThrow();
+        CodeElement prev = null;
+        for (CodeElement curr : cinit.code().get().elementList()) {
+            if (curr instanceof FieldInstruction fi && fi.opcode() == Opcode.PUTSTATIC
+                    && fi.owner().equals(cm.thisClass()) && fi.name().stringValue().endsWith("$op")) {
+                var lci = (ConstantInstruction.LoadConstantInstruction) prev;
+                var pe = (StringEntry) lci.constantEntry();
+                res.add(new OpFieldAndIR(fi.name().stringValue().substring(fi.name().stringValue().indexOf(':') + 2), pe.stringValue()));
```

Comment:
Suggest you return pairs of `FieldRefEntry` (or `FieldInstruction`)  and `String`, thereby you retain the full field name for possible reduction later on if needed, and you can reuse later when removing them.

Was there an issue with using the synthetic field name as a synthetic method name?

Reply:
For method name, I had to remove the class part.